pkg/kn/commands/route/route_describe_test.go

```diff
@@ -0,0 +1,130 @@
+	knParams := &commands.KnParams{}
+	cmd, fakeRoute, buf := commands.CreateTestKnCommand(NewRouteCommand(knParams), knParams)
+	fakeRoute.AddReactor("*", "*",
+		func(a client_testing.Action) (bool, runtime.Object, error) {
+			action = a
+			return true, response, nil
+		})
+	cmd.SetArgs(args)
+	err = cmd.Execute()
+	if err != nil {
+		return
+	}
+	output = buf.String()
+	return
+}
+
+func TestCompletion(t *testing.T) {
+	var expectedRoute v1alpha1.Route
+
+	setup := func(t *testing.T) {
+		expectedRoute = v1alpha1.Route{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Route",
+				APIVersion: "knative.dev/v1alpha1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "foo",
+				Namespace: "default",
+			},
+		}
+	}
+
+	t.Run("requires the route name", func(t *testing.T) {
+		_, _, err := fakeRouteDescribe([]string{"route", "describe"}, &v1alpha1.Route{})
+		assert.Assert(t, err != nil)
+		assert.Assert(t, strings.Contains(err.Error(), "requires the route name."))
+	})
+
+	t.Run("describe a valid route with default output", func(t *testing.T) {
+		setup(t)
+
+		action, output, err := fakeRouteDescribe([]string{"route", "describe", "test-foo"}, &expectedRoute)
+		assert.Assert(t, err == nil)
+		assert.Assert(t, action != nil)
+		assert.Assert(t, action.Matches("get", "routes"))
+
+		jsonData, err := yaml.YAMLToJSON([]byte(output))
+		assert.Assert(t, err == nil)
+
+		var returnedRoute v1alpha1.Route
+		err = json.Unmarshal(jsonData, &returnedRoute)
+		assert.Assert(t, err == nil)
+		assert.Assert(t, equality.Semantic.DeepEqual(expectedRoute, returnedRoute))
+	})
+
+	t.Run("describe a valid route with special output", func(t *testing.T) {
+		t.Run("describe a valid route with YAML output", func(t *testing.T) {
+			setup(t)
+
+			action, output, err := fakeRouteDescribe([]string{"route", "describe", "test-foo", "-oyaml"}, &expectedRoute)
+			assert.Assert(t, err == nil)
+			assert.Assert(t, action != nil)
+			assert.Assert(t, action.Matches("get", "routes"))
+
+			jsonData, err := yaml.YAMLToJSON([]byte(output))
+			assert.Assert(t, err == nil)
+
+			var returnedRoute v1alpha1.Route
+			err = json.Unmarshal(jsonData, &returnedRoute)
+			assert.Assert(t, err == nil)
+			assert.Assert(t, equality.Semantic.DeepEqual(expectedRoute, returnedRoute))
+		})
+
+		t.Run("describe a valid route with JSON output", func(t *testing.T) {
+			setup(t)
+
+			action, output, err := fakeRouteDescribe([]string{"route", "describe", "test-foo", "-ojson"}, &expectedRoute)
+			assert.Assert(t, err == nil)
+			assert.Assert(t, action != nil)
+			assert.Assert(t, action.Matches("get", "routes"))
+
+			var returnedRoute v1alpha1.Route
+			err = json.Unmarshal([]byte(output), &returnedRoute)
+			assert.Assert(t, err == nil)
+			assert.Assert(t, equality.Semantic.DeepEqual(expectedRoute, returnedRoute))
+		})
+
+		t.Run("describe a valid route with name output", func(t *testing.T) {
+			setup(t)
+
```

Comment:
Are there a fix list for what `name` can be? If yes then IMO code should guard on these, unless APIs gives errors and you can return error with a nice message to user that the `-oname` they used is invalid...

Reply:
the `-oname` returns the `resource tpye`/`resource name`, so I checked if its result contains the resource name directly.

Reply:
What is the benefit of `-oname` when you already have to provide the name as mandatory argument ?

Reply:
the `-oname` is another format type to only show the resource name. I just add as anoter format type test. Like this:
```
jordanzt@zhangtaodembp:~/Works/workspaces/gopath/src/github.com/knative/client$ go run cmd/kn/main.go route describe helloworld -oname
route.serving.knative.dev/helloworld
```

Reply:
Still I believe that -oname is its a no-op here. Why would you call this command if you already know the name + group. Agreed, you could check for whether a revision exists or not. So I'm not totally against it and we might adapt this when route describe changes anyway. So I'm fine with it for now.